### ISSUE TITLE
[KMCompiler][NPU] Incorrect blockNum causes redundant Triton program execution

### ISSUE THREAD

**Issue by Caeruleann** [flagos-ai/libtriton_jit#51]
# libtriton_jit NPU backend: incorrect blockNum causes redundant Triton program execution

## Environment

| Component | Version |
|---|---|
| Hardware | Ascend 910B4-1 |
| CANN | 8.5.0, npu-smi 25.2.0 |
| triton (triton-ascend) | 3.2.0 (flagtree 0.5.0+ascend3.2) |
| torch_npu | 2.6.0rc1 |
| libtriton_jit | commit `50a4259` |
| OS / Python | Ubuntu 22.04.5, aarch64, Python 3.11.13 |

**Conclusion:** libtriton_jit passes the Triton logical grid size as `blockNum` to `rtKernelLaunch` without limiting it to the available AIV parallel blocks, unlike the triton-ascend generated launcher.

---

## Reproduction (`test.py`)

```python
import torch
import flag_gems.c_operators as co
import flag_gems
x1 = torch.ones(10000000, dtype=torch.bfloat16, device="npu")
sum_co = co.sum(x1, None)        # C++ wrapper  (libtriton_jit)
sum_py = flag_gems.sum(x1)       # Python wrapper (triton-ascend)
print(torch.equal(sum_co, sum_py))
```

Both paths launch the **same** `sum_kernel_1` with `grid = ⌈10M / 4096⌉ = 2442` (≫ 40), because both reuse the tile formula in `ops/sum.py`.

## Instrumentation used to measure pid repetition

To count how many times each pid is executed, three temporary changes were made.
They do not alter the result because `sum_kernel_1` is idempotent under
re-execution: each pid deterministically writes `mid[pid] = partial_sum`.
This issue is partially hidden by this property. Kernels with atomic updates or
other non-idempotent side effects may produce incorrect results instead of only
suffering performance degradation.

**(1) Triton kernel** — `src/flag_gems/ops/sum.py`, `sum_kernel_1` (added a `counter` arg + one `atomic_add`):

```python
def sum_kernel_1(inp, mid, M, counter, BLOCK_SIZE: tl.constexpr):
    ...
    pid = ext.program_id(0)
    tl.atomic_add(counter + pid, 1)          # +1 every time this pid runs
    offset = pid * BLOCK_SIZE + tl.arange(0, BLOCK_SIZE)
    ...
```

**(2) Python wrapper** — `ops/sum.py`, `sum()` (allocate counter, sync, print):

```python
counter = torch.zeros((mid_size,), dtype=torch.int32, device=inp.device)
with torch_device_fn.device(inp.device):
    sum_kernel_1[(mid_size, 1, 1)](inp, mid, M, counter, block_size)
    sum_kernel_2[(1, 1, 1)](mid, out, mid_size, block_mid)
print(f"[PY ] sum_kernel_1: grid={mid_size} distinct_pids={(counter > 0).sum().item()} "
      f"total_exec={counter.sum().item()} max_hits={counter.max().item()}")
```

**(3) C++ wrapper** — `lib/sum.cpp`, `sum()` (allocate counter, sync, print):

```cpp
at::Tensor counter = torch::zeros({mid_size}, self.options().dtype(at::kInt));
...
sum_kernel_1(raw_stream, mid_size, 1, 1, num_warps, num_stages, self, mid, M, counter, block_size);
std::cerr << "[C++] sum_kernel_1: grid=" << mid_size
          << " distinct_pids=" << (counter > 0).to(at::kInt).sum().item<int>()
          << " total_exec=" << counter.sum().item<int>()
          << " max_hits=" << counter.max().item<int>() << "\n";
```

## Execution results

```
$ python test.py

[C++] sum_kernel_1: grid=2442 distinct_pids=2442 total_exec=151404 max_hits=75641
[PY ] sum_kernel_1: grid=2442 distinct_pids=2442 total_exec=2480  max_hits=39
True
```

msprof device `Task Duration` of `sum_kernel_1` (clean kernel, avg of 70 launches, grid=2442):

| Path | blockNum → rtKernelLaunch | sum_kernel_1 device time | result |
|---|---|---|---|
| Python wrapper (triton-ascend) | 40 | **36.0 µs** | correct |
| C++ wrapper, **(bug)** | 2442 (= grid) | **3914.5 µs** | correct |

## Cause and effect

**(cause):** libtriton_jit's `npu_backend.h::launch_kernel` passes `grid_x*grid_y*grid_z` as `blockNum` to `rtKernelLaunch` without capping it. On Ascend, `blockNum` should be limited by the available AIV blocks. Otherwise, Triton programs may be executed repeatedly.

**(effect):** the C++ wrapper does **61× more pid-executions** (151 404 vs 2480) and is **109× slower** (3914 µs vs 36 µs) than the Python wrapper, for the *identical* kernel, grid, input, and result (`True`). 

**Why the Python wrapper is unaffected:** triton-ascend's launcher caps `blockNum` to the available AIV parallel blocks. The launcher triton-ascend actually generates was inspected with `TRITON_DEBUG=1` (dumped as `launcher_cxx11abi1.cxx`); immediately before `rtKernelLaunch` it contains:

```c
uint32_t blockNum = gridX * gridY * gridZ;     // = grid (2442)
...
blockNum = std::min(blockNum, (uint32_t)40);   // caps to available AIV parallel blocks
ret = rtKernelLaunch(func, blockNum, ...);
```

So the Python wrapper passes **40**, not 2442, to `rtKernelLaunch` — identical to the capped C++ path (which is why both print `total_exec=2480`). libtriton_jit's `npu_backend.h` omits this `std::min`, passes 2442, and re-execution. 

## triton-ascend behavior (the reference)

triton-ascend already caps `blockNum` — `driver.py:518` queries the core count and `driver.py:765` emits `blockNum = std::min(blockNum, num_physical_blocks)`:

```python
# driver.py:518
num_physical_blocks = npu_utils.get_aivector_core_num()   # == rtGetAiCoreCount() * 2 = 40 on 910B
```
```c
// driver.py:765 (gated on TRITON_ALL_BLOCKS_PARALLEL)
blockNum = std::min(blockNum, (uint32_t)num_physical_blocks);
ret = rtKernelLaunch(func, blockNum, ...);                 // driver.py:711
```

## Fix

I mirror exactly what triton-ascend's generated launcher does. That launcher is inspectable via `TRITON_DEBUG=1` (dumped as `launcher_cxx11abi1.cxx`); its launch path caps `blockNum` to the core count and runs `rtKernelLaunch` through `at_npu::native::OpCommand`:

```c
// from the triton-ascend generated launcher (launcher_cxx11abi1.cxx)
auto launch_call = [=]() -> rtError_t {
    uint32_t blockNum = gridX * gridY * gridZ;
    blockNum = std::min(blockNum, (uint32_t)40);              // ← the cap
    ...
    ret = rtKernelLaunch(func, blockNum, static_cast<void*>(&args), sizeof(args), NULL, stream);
    return ret;
};
at_npu::native::OpCommand cmd;
cmd.Name(name.c_str()).SetCustomHandler(launch_call).Run();   // ← launch via OpCommand
```

Applying the same to libtriton_jit's `include/triton_jit/backends/npu_backend.h::launch_kernel` — for now, the direct cap as in the dump, plus the OpCommand-based launch:

```cpp
// cap blockNum to the AIV core count, exactly as the triton-ascend launcher does
blockNum = std::min(blockNum, (uint32_t)40);

// launch via OpCommand (as in the dump) — wraps rtKernelLaunch in a handler run
// The launch path should follow the triton-ascend generated launcher.
auto launch_call = [=]() mutable -> rtError_t {
    return rtKernelLaunch(kernel, blockNum, arg_buffer.data(),
                          static_cast<uint32_t>(arg_buffer.size()), nullptr, stream);
};
at_npu::native::OpCommand cmd;
cmd.Name("triton_kernel").SetCustomHandler(launch_call).Run();
```

It can be queried dynamically via `rtGetAiCoreCount()` to adapt to other Ascend chips. The temporary workaround is to cap it to 40 on 910B.

## Collateral issue: libtriton_jit lacks msprof instrumentation.

triton-ascend's generated launcher reports rich per-kernel metadata to msprof via `MsprofReportCompactInfo` / `MsprofReportAdditionalInfo` (op name, task type, **blockDim**, and input/output tensor shapes & dtypes — all visible in the `TRITON_DEBUG=1` dump). libtriton_jit's `npu_backend.h` does none of this — it only calls `rtKernelLaunch(...)`, so msprof receives none of that metadata.

This is visible directly when profiling `test.py`. In the same `op_summary.csv`, the two `sum_kernel_1` launches differ sharply:

| msprof field | C++ wrapper (`sum_kernel_1_0`) | Python wrapper (`sum_kernel_1`) |
|---|---|---|
| Block Dim | **0** (missing) | **40** |
| Input Shapes | N/A | `"10000000"` |
| Input Data Types | N/A | `DT_BF16` |
| Output Shapes | N/A | `"2442"` |
| aiv_time | 0.0 | populated |

So libtriton_jit-launched kernel shows **Block Dim = 0**, no tensor shapes/types, and no aiv timing in msprof, which are exactly the fields triton-ascend populates.
